### PR TITLE
 Migrate test-integration-cli experimental ipvlan test to integration 

### PR DESCRIPTION
**Pull request by vdemeester**
<del>Build on top of #36697, only the second commit (007b7f7) should be reviewed :angel: </del>

All `Ipvlan` related test on `DockerSuite` and `DockerNetworkSuite`
are migrated to `ipvlan_test.go`.

The end goal being to remove the `experimental` builds.

It also refactors macvlan tests a bit (quicker) *and* it put `macvlan` and `ipvlan` tests on there own folder in order to be isolated and have quick "suite).

cc @fcrisciani @selansen 

:lion: 

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

**Comment by vdemeester** [comment]
Updated to do a bit more than just `ipvlan` migration. It also refactors macvlan tests a bit (quicker) *and* it put `macvlan` and `ipvlan` tests on there own folder in order to be isolated and have quick "suite).

**Comment by fcrisciani** [comment]
@ctelfer can you take a look too?

**Review by fcrisciani — APPROVED** [review]
high level I guess looks good

**Review by ctelfer — COMMENTED** [review]
Sorry it has taken a while to get through the code.  LGTM overall.  Just one comment about a cleanup function.

File: integration/network/helpers.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	// ip link add link <master> name <master>.<VID> type vlan id <VID>
+	icmd.RunCommand("ip", "link", "add", "link", master, "name", slave, "type", "vlan", "id", id).Assert(t, icmd.Success)
+	// ip link set <sub_interface_name> up
+	icmd.RunCommand("ip", "link", "set", slave, "up").Assert(t, icmd.Success)
+}
+
+// DeleteInterface deletes a network interface
+func DeleteInterface(t *testing.T, ifName string) {
+	icmd.RunCommand("ip", "link", "delete", ifName).Assert(t, icmd.Success)
+	icmd.RunCommand("iptables", "-t", "nat", "--flush").Assert(t, icmd.Success)
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply integration/network/helpers.go]: @ctelfer oh interesting… I took what was previously here so the behavior doesn't change from previous tests. cc @mavenugo to get more insight :angel:

File: integration/network/ipvlan_test.go [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+	"github.com/docker/docker/integration-cli/daemon"
+	"github.com/docker/docker/integration/internal/container"
+	"github.com/gotestyourself/gotestyourself/assert"
+	"github.com/gotestyourself/gotestyourself/skip"
+	"golang.org/x/net/context"
+)
+
+func TestDockerNetworkIpvlanPersistance(t *testing.T) {
+	// verify the driver automatically provisions the 802.1q link (di-dummy0.70)
+	skip.If(t, testEnv.DaemonInfo.OSType != "linux")
```
Comment (thaJeztah) [inline integration/network/ipvlan_test.go]: Commented on another PR; we should go over the tests and be consistent (either "!= linux" or "== windows").

No need to fix for this PR, as it's a general thing to look at

File: integration/network/macvlan_test.go [hunk 3]
```diff
@@ -189,7 +192,7 @@ func TestDockerNetworkMacvlanMultiSubnet(t *testing.T) {
 	skip.If(t, testEnv.DaemonInfo.OSType != "linux")
 	skip.If(t, testEnv.IsRemoteDaemon())
 	skip.If(t, !macvlanKernelSupport(), "Kernel doesn't support macvlan")
-	t.Skip("Temporarily skipping while investigating sporadic v6 CI issues")
+	// t.Skip("Temporarily skipping while investigating sporadic v6 CI issues")
```
Comment (thaJeztah) [inline integration/network/macvlan_test.go]: Guess this can be removed

File: integration/network/macvlan_test.go [hunk 4]
```diff
@@ -48,292 +49,275 @@ func TestDockerNetworkMacvlanPersistance(t *testing.T) {
...
+		ctx := context.Background()
+		id1 := container.Run(t, ctx, client,
+			container.WithNetworkMode("dualstackbridge"),
+		)
+
+		// Validate macvlan bridge mode defaults gateway sets the default IPAM next-hop inferred from the subnet
+		result, err := container.Exec(ctx, client, id1, []string{"ip", "route"})
+		assert.NilError(t, err)
+		assert.Check(t, strings.Contains(result.Combined(), "default via 172.28.130.1 dev eth0"))
+		// Validate macvlan bridge mode sets the v6 gateway to the user specified default gateway/next-hop
```
Comment (thaJeztah) [inline integration/network/macvlan_test.go]: Wondering if we should start moving comments like this as custom message (not for this PR)

File: integration/network/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply integration/network/helpers.go]: As discussed on slack: these tests set up a new daemon, so this shouldn't be a problem, but perhaps we can add a TODO to these (as a follow-up)

File: integration/network/helpers.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			if network.Name == name {
+				return cmp.ResultFailure(fmt.Sprintf("network %s is still present", name))
+			}
+		}
+		return cmp.ResultSuccess
+	}
+}
+
+// CheckKernelMajorVersionGreaterOrEqualThen returns whether the kernel version is greater or equal than the one provided
+func CheckKernelMajorVersionGreaterOrEqualThen(kernelVersion int, majorVersion int) bool {
```
Comment (thaJeztah) [inline integration/network/helpers.go]: Should probably be moved somewhere else, as it's not network-specific (may even be another helper like this already)

**Review by thaJeztah — APPROVED** [review]
LGTM, left some nits/questions but it's ok to address those separately

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
And a conflict now.

**Comment by vdemeester** [comment]
Gonna fix the conflicts and take @thaJeztah comments into account tomorrow morning (Paris time :wink:)

**Comment by vdemeester** [comment]
Rebased :angel:

File: integration/network/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cpuguy83) [reply integration/network/helpers.go]: The problem is these daemons aren't in their own network namespace and it's flushing everything in the main namespace.

**Pull request merged**